src/webgpu/api/validation/vertex_access.spec.ts

```diff
@@ -47,45 +59,318 @@ TODO: make sure this isn't already covered somewhere else, review, organize, and
+import { kVertexFormatInfo } from '../../capability_info.js';
+import { GPUConst } from '../../constants.js';
 
 import { ValidationTest } from './validation_test.js';
 
-export const g = makeTestGroup(ValidationTest);
+// Class that indicate how to call a draw function
+// TODO: implement this class to hold the draw call parameter and insert draw call to encoder
+class DrawCall {
+  callDraw(encoder: GPURenderEncoderBase) {
+    unreachable();
+  }
+}
+
+// Interfaces that indicate how to call setIndexBuffer and setIndexBuffer
+interface SetBufferParam {
+  buffer: GPUBuffer;
+  offset?: number;
+  size?: number;
+}
+
+interface SetVertexBufferParam extends SetBufferParam {
+  slot: number;
+}
+
+interface SetIndexBufferParam extends SetBufferParam {
+  indexFormat: GPUIndexFormat;
+}
+
+// Class that indicate how to make input attribute for vertex shader module
+interface VertexShaderInput {
+  shaderLocation: GPUIndex32;
+  wgslType: string;
+}
+
+class F extends ValidationTest {
+  // Create a vertex shader module with given attributes wgsl type
+  generateVertexShaderModuleFromInputDescriptor(
+    vertexInputDescriptor: VertexShaderInput[]
+  ): GPUShaderModule {
+    const shaderInput = vertexInputDescriptor
+      .map((attr, index) => `[[location(${attr.shaderLocation})]] var_${index} : ${attr.wgslType}`)
+      .join(', ');
+
+    const code = `[[stage(vertex)]]
+    fn main(${shaderInput}) -> [[builtin(position)]] vec4<f32> {
+        return vec4<f32>(0.0, 1.0, 0.0, 1.0);
+    }`;
+
+    return this.device.createShaderModule({ code });
+  }
+
+  // Create a vertex shader module from vertex buffer layout array, typically from the pipeline
+  // descriptor, and use the perfect match wgsl type, e.g. vec4<f32> for float32x4 for all attributes
+  generateVertexShaderModuleFromBufferLayout(
+    vertexBufferLayouts: GPUVertexBufferLayout[]
+  ): GPUShaderModule {
+    return this.generateVertexShaderModuleFromInputDescriptor(
+      vertexBufferLayouts
+        .reduce<GPUVertexAttribute[]>((a, c) => [...a, ...c.attributes], [])
+        .map(
+          attr =>
+            ({
+              shaderLocation: attr.shaderLocation,
+              wgslType: kVertexFormatInfo[attr.format].wgslType,
+            } as const)
+        )
+    );
+  }
+
+  /**
+   * Create a render pipeline with given vertex buffer layouts. The vertex shader module is created
+   * using the exact matching WGSL type (as in typeInfoMap) for all attributes in all buffers in
+   * layouts. If a different vertex shader is wanted (especially when testing WGSL type
+   * compatibility), a standalone list of vertex shader input descriptor should be given.
+   * @param vertexBufferLayouts A list of GPUVertexBufferLayout to be used in render pipeline descriptor,
+   * and also used to generate the vertex shader module if `vertexShaderInputs` is not given.
+   * @param vertexShaderInputs Optional, the standalone list of vertex shader input descriptor.
+   * @returns
```

Comment:
(fyi `@param` and `@returns` are not strictly necessary, we mainly want to use jsdoc so the docs appear in the documentation generator (not relevant in .spec.ts files anyway) and can be seen in IDEs when hovering functions.)

Reply:
Thanks!